cmd/fs-v1-multipart.go

```diff
@@ -334,36 +430,42 @@ func (fs fsObjects) PutObjectPart(bucket, object, uploadID string, partID int, s
 	postUploadIDLock.Lock()
 	defer postUploadIDLock.Unlock()
 
-	// Just check if the uploadID exists to avoid copy if it doesn't.
+	// Just check if the uploadID exists to not proceed further,  if it doesn't.
 	if !fs.isUploadIDExists(bucket, object, uploadID) {
 		return "", traceError(InvalidUploadID{UploadID: uploadID})
 	}
 
-	fsMetaPath := pathJoin(uploadIDPath, fsMetaJSONFile)
-	fsMeta, err := readFSMetadata(fs.storage, minioMetaMultipartBucket, fsMetaPath)
+	fsMetaPath := pathJoin(fs.fsPath, minioMetaMultipartBucket, uploadIDPath, fsMetaJSONFile)
+	metaFile, err := fs.rwPool.Create(fsMetaPath)
+	if err != nil {
+		return "", toObjectErr(traceError(err), bucket, object)
+	}
+	defer metaFile.Close()
+	fsMeta := fsMetaV1{}
+	_, err = fsMeta.ReadFrom(io.NewSectionReader(metaFile, 0, metaFile.Size()))
 	if err != nil {
 		return "", toObjectErr(err, minioMetaMultipartBucket, fsMetaPath)
 	}
 	fsMeta.AddObjectPart(partID, partSuffix, newMD5Hex, size)
 
-	partPath := path.Join(bucket, object, uploadID, partSuffix)
+	partPath := pathJoin(bucket, object, uploadID, partSuffix)
 	// Lock the part so that another part upload with same part-number gets blocked
 	// while the part is getting appended in the background.
 	partLock := globalNSMutex.NewNSLock(minioMetaMultipartBucket, partPath)
 	partLock.Lock()
-	err = fs.storage.RenameFile(minioMetaTmpBucket, tmpPartPath, minioMetaMultipartBucket, partPath)
-	if err != nil {
+
+	fsNSPartPath := pathJoin(fs.fsPath, minioMetaMultipartBucket, partPath)
+	if err = fsRenameFile(fsPartPath, fsNSPartPath); err != nil {
 		partLock.Unlock()
-		return "", toObjectErr(traceError(err), minioMetaMultipartBucket, partPath)
+		return "", toObjectErr(err, minioMetaMultipartBucket, partPath)
 	}
-	uploadIDPath = path.Join(bucket, object, uploadID)
-	if err = writeFSMetadata(fs.storage, minioMetaMultipartBucket, path.Join(uploadIDPath, fsMetaJSONFile), fsMeta); err != nil {
+	if _, err = fsMeta.WriteTo(metaFile); err != nil {
 		partLock.Unlock()
 		return "", toObjectErr(err, minioMetaMultipartBucket, uploadIDPath)
 	}
```

Comment:
there is a corner case where the part gets replaced by another minio server before `os.OpenFile(preparePath(tmpObjPath), os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)` happens in appendPart(). In which case the minio server will append a different part but the ETAG that it remembers will not be the md5sum of the part that it appended.

i think this can be fixed in a different patch as it will complicate this patch